src/auth/identitypoolclient.ts

```diff
@@ -0,0 +1,151 @@
+    };
+  };
+}
+
+/**
+ * Defines the Url-sourced and file-sourced external account clients mainly
+ * used for K8s and Azure workloads.
+ */
+export class IdentityPoolClient extends BaseExternalAccountClient {
+  private readonly file?: string;
+  private readonly url?: string;
+  private readonly headers?: {[key: string]: string};
+
+  /**
+   * Instantiate an IdentityPoolClient instance using the provided JSON
+   * object loaded from an external account credentials file.
+   * An error is thrown if the credential is not a valid file-sourced or
+   * url-sourced credential.
+   * @param options The external account options object typically loaded
+   *   from the external account JSON credential file.
+   * @param additionalOptions Optional additional behavior customization
+   *   options. These currently customize expiration threshold time and
+   *   whether to retry on 401/403 API request errors.
+   */
+  constructor(
+    options: IdentityPoolClientOptions,
+    additionalOptions?: RefreshOptions
+  ) {
+    super(options, additionalOptions);
+    this.file = options.credential_source.file;
+    this.url = options.credential_source.url;
+    this.headers = options.credential_source.headers;
+    if (!this.file && !this.url) {
+      throw new Error('No valid Identity Pool "credential_source" provided');
+    }
+  }
+
+  /**
+   * Triggered when a external subject token is needed to be exchanged for a GCP
+   * access token via GCP STS endpoint.
+   * This uses the `options.credential_source` object to figure out how
+   * to retrieve the token using the current environment. In this case,
+   * this either retrieves the local credential from a file location (k8s
+   * workload) or by sending a GET request to a local metadata server (Azure
+   * workloads).
+   * @return A promise that resolves with the external subject token.
+   */
+  async retrieveSubjectToken(): Promise<string> {
+    if (this.file) {
+      return await this.getTokenFromFile(this.file!);
+    } else {
+      return await this.getTokenFromUrl(this.url!, this.headers);
+    }
+  }
+
+  /**
+   * Looks up the external subject token in the file path provided and
+   * resolves with that token.
+   * @param file The file path where the external credential is located.
+   * @return A promise that resolves with the external subject token.
+   */
+  private getTokenFromFile(filePath: string): Promise<string> {
+    // Make sure there is a file at the path. lstatSync will throw if there is
+    // nothing there.
+    try {
+      // Resolve path to actual file in case of symlink. Expect a thrown error
+      // if not resolvable.
+      filePath = fs.realpathSync(filePath);
+
+      if (!fs.lstatSync(filePath).isFile()) {
+        throw new Error();
+      }
+    } catch (err) {
+      err.message = `The file at ${filePath} does not exist, or it is not a file. ${err.message}`;
+      throw err;
+    }
+
+    const stream = fs.createReadStream(filePath);
```

Comment:
an alternate approach to `createReadStream`, I believe would be using the promises API:

```
const {readFile} = require('fs').promises

async function main () {
  return readFile(filePath)
}
```

Reply:
I ended up using `utils.promisify(fs.readFile)` since Node 10 does not seem to play well with `fs.promises`.